### DISCUSSION TITLE
Discussion of a new manifest

### DISCUSSION THREAD

**Discussion by SteveLasker** [opencontainers/artifacts#41] [General]
OCI Artifacts [PR #29](https://github.com/opencontainers/artifacts/pull/29) and [PR #37](https://github.com/opencontainers/artifacts/pull/37) both recommend a new manifest be used for distribution of new artifact types. The following captures the discussions that led to the recommendation for a new manifest type:

1. **The `oci.image.manifest` is specific to the container image format**. While it made sense at the time, distribution has evolved to support a wide range of artifact types including Helm, Singularity, CNAB, OPA, WASM, VMs, and many other types.  
How do the OCI specs refactor the needs of one artifact type from all other artifact types?
1. **`config` REQUIRED**: the [image-spec declares config as REQUIRED](https://github.com/opencontainers/image-spec/blob/master/manifest.md#image-manifest-property-descriptions). Most artifact types don't need a config object. To work around this, OCI Artifacts promotes pushing a null config. This has surfaced as a challenge for some registries to support.  
Would the image-spec change `config` to OPTIONAL?
1. **`layers` are REQUIRED** and [specified as ordinal](https://github.com/opencontainers/image-spec/blob/master/manifest.md#image-manifest-property-descriptions). Most other artifact types store collections of files, but are not ordinal. Some artifacts may not have any layers/blobs. Further, other reference types may persist information as annotations, enhancing an existing artifact without adding a layer/blob.  
Would the image-spec change `layers` to OPTIONAL?
1. **Differentiating artifact types**: Initially, OCI Artifacts attempted to add an `artifactType` property to the image-spec, but the feedback was to not add new properties to the versioned spec, as it risked breaking existing clients. Instead, the working group *"hacked"* in the concept by (overloading) the `manifest.config.mediaType`. This may be a bit confusing as most non-image artifacts do not have a config.  
See: [Defining OCI Artifact Types](https://github.com/opencontainers/artifacts/blob/master/artifact-authors.md#defining-oci-artifact-types) for more info.
1. **Lack of versioning support**: [PR #37](https://github.com/opencontainers/artifacts/pull/37) goes into the details. Long term, as captured in PR #37, registries and artifact types could benefit from a new generic manifest format, supporting these artifact reference type scenarios and more. Some examples:  
    - [Added encryption mediatype doc #15](https://github.com/opencontainers/artifacts/pull/15)
    - Multi-arch complexities of using image-index
    - [extensible image format](https://hackmd.io/@cyphar/ociv2-brainstorm#Extensibility) as one of the many [Image v2 proposals](https://hackmd.io/@cyphar/ociv2-brainstorm)
1. ** Non-container image tooling must refer to the image-spec**. [ORAS](https://github.com/deislabs/oras) is one of several efforts to provide generalized libraries for utilizing an **O**CI **R**egistry **A**s **S**torage. While ORAS focuses on all non-image artifacts, it winds up pulling in references to oci-image and containerd as many of the [golang spec files](https://github.com/opencontainers/image-spec/tree/master/specs-go) are within these projects. Reviewing the [Artifacts Scope](https://github.com/opencontainers/artifacts/blob/master/artifact-authors.md#scope), the vision was to enable and support artifact-specific specs to evolve. With some refactoring, Artifact types that use the ORAS libraries can reference OCI Artifacts, and evolve beyond certain OCI image-spec v1 limitations.  
_**Note:** Phase 1 is NOT intended to support OCI image types. Phase 2, with the evolution of PR #37, is when phase 1 reference types and other artifacts *could* move/migrate._

After working through all the above reasons for creating a new manifest for the existing issue, consider:

**Reference Type support:** How to enable a wide range of new scenarios for adding information to a registry, including Signature(s), SBoM, Security Scan Results, meta-data, gpl source that must extend (reference) existing types? While supporting [lifecycle management](https://github.com/SteveLasker/artifacts/blob/oci-artifact-manifest/artifact-manifest.md#lifetime-management) to know how and when to delete reference artifacts

## Considerations for adding to the existing manifests

Various discussions did focus on using the OCI image-spec. Whether it had (required) a new version, or would rely on the existing version.

### Why not add to the existing image-spec?

**Separation of concerns**: #1 above. Even if the image-spec could solve all the problems, it's still the container image-spec, which creates a requirement on the image-spec maintainers to account for all other artifact types, and similar scenarios. By listing certain requirements for each new artifact type, the distribution and artifact specs may/can then consider implementing a generic approach to cover the new requirements. For instance, [making a config object just one of the descriptors in the `[blobs]` collection](https://github.com/SteveLasker/artifacts/blob/e17c1874fd477e435c158e6beed8c166fd32190c/manifest-examples/image.json), providing the container image artifact type to have the desired behavior, but not having to support a REQUIRED or even OPTIONAL property in the schema. The container image clients *could* look for a single config `mediaType` and process it. 

**NOTE:**, *Phase 1 isn't proposing the image-spec move to PR#29. This is an example of how a future generic artifact manifest (PR #37) could support artifact-specific scenarios.*

### Why can't optional properties be added to the image-spec

**It can, but what does optional imply?** There's a pretty big difference between adding an optional property like an annotation that registries largely ignore and something that impacts GC and lifecycle management. The purpose of a spec is to provide confidence in consistency to those that support a specific version of the spec. If 1.0 image-spec manifests PUSHED to a registry *sometimes* had `subjectManifest`, how would a user understand the expected behavior from a registry that accepted it because the registry didn't limit additional schema elements? How would that compare with registries that support the reference type concept? How would users get the artifacts out of the registry without the `/referrers` api?

### Garbage collection of untagged manifests

Registries implement GC differently, however many have the concept of untagged manifests are garbage collected. Docker Hub will maintain any digest that was once associated with a tag. Associated in the image-spec means the digest was tagged. ACR & ECR support automatically deleting untagged manifests, which many customers have on by default as it allows them to manage the sizes of their registries. The reference types are considered enhancements to an artifact, and not thought to have a life unto themselves, therefore they wouldn't be tagged.

In the image-spec case, a registry may accept an `oci.image.manifest` with the `subjectManifest`, but without GC implementations to understand the reference, the artifact would be automatically deleted at some later point.

While clients could tag the reference types, this only adds to the ambiguity of content lifecycle management and how users might delete the related artifacts when the `subjectManifest` is deleted.

_**NOTE:** See [Content Discovery for more info](https://github.com/SteveLasker/artifacts/blob/oci-artifact-manifest/artifact-manifest.md#content-discovery)_

### Timed release for the fall of 2021

The Notary v2 effort has a time-bounded/quality-focused release and needs to ship a core set of capabilities. The work has identified a larger set of usecases that benefit from reference types, including SBoMs, Scan Results. Because the effort is time & quality bound, the working group knows they'll intentionally miss certain features and not have a chance to incorporate some things that will surface along the way. By defining a different manifest format, registry ingestion can isolate the handling of this specific versioned release. If the properties merged into the image-spec, ingestion would forever be checking for the existence of these elements. As registries move to RC2 and v1 of the manifest, this codepath is isolated and *eventually* retired. *(recognizing it's hard to deprecate content, but all the more important to have this codepath isolated)*

### Why create a new manifest now, vs. wait for PR #37 to complete?

This comes back to how should registries process manifests with data that must be acted upon. To provide any reasonable performance, registries will want to reverse index the `subjectManifest`, likely with the `referenceType` property to provide filtered results in the `/referrers` api.

The underlying schema of the phase 1 manifest will change as it evolves to PR #37. If the above parsing logic is isolated using the distinct type as a switch, registries can isolate the impacts in execution/implementation.

### Image toolchain isolations

Container image toolchains know to pull images by tags and sometimes by digests. However, they also know how to parse specific manifest types (`image.manifest`, `image.index`). By isolating this new artifact reference type, image clients are able to parse it with separate logic and thus avoid partial failure scenarios.

### Should image toolchains know about signatures

Notary v2, refers to this as the Trojan Horse Protection. Before the image toolchains of containerd, Kubernetes and every other container runtime even pulls the image, some sort of admissions controller is implemented. k8s may use OPA/Gatekeeper, while non k8s implementations may use a containerd plug-in. Within the admission controller, the signature is discovered, pulled, and verified. Only if it passes does the pipeline continue, handing the image pull url to the container runtime. This separation of concerns enables signature verification clients to focus on their artifact type while the container runtime goes unchanged.

### Verification implementations

To test if a new manifest model would even work, several end-to-end verification implementations were created. [Notary v2 Prototype 1](https://github.com/notaryproject/nv2/blob/prototype-1/docs/nv2/demo-script.md) tested a `references` api from CNCF distribution that returned a list of descriptors, while the current prototype returns a collection of referenced manifests to make the desired artifact pulls efficient.

The current implementations are:

- [CNCF Distribution](https://github.com/notaryproject/distribution/tree/prototype-2)
- [ORAS experimental test client](https://github.com/deislabs/oras/blob/prototype-2/docs/artifact-manifest.md)
- [Notary v2 test client](https://github.com/notaryproject/nv2/blob/prototype-2/docs/nv2/demo-script.md)

## Call for early adopters

While cloud providers and vendors could quickly move forward with a vendor-specific solution, a vendor-neutral approach, where artifacts and their references can move across vendor-provided registries, is a better starting point. Some registry products and projects may decide to wait for a more stable solution.
However, those that wish to implement secure supply chain scenarios within registries sooner would have a vendor-neutral design to leverage, get feedback and help secure the cloud-native supply chain of artifacts.

**Comment by imjasonh** [comment]
> The `oci.image.manifest` is specific to the container image format.

I think that's up to interpretation. Plenty of folks are clearly successfully using the format to store other stuff, otherwise it's just a small misnomer in the content type, that no user really sees anyway.

> Would the image-spec change config to OPTIONAL?
> Would the image-spec change layers to OPTIONAL?

Sure, why not? If there's use cases and sufficiently motivated people to push for the change, I think that could be great. I assume it wouldn't be a gigantic change for registries to loosen validation for incoming requests, at least not as gigantic as supporting a whole new artifact type.

> Lack of versioning support

Versioning is important for negotiating and staging breaking changes, but if spec authors are careful not to add breaking changes (requiring a new field, changing the type of a field, inventing and requiring new types), then I believe a sufficiently diligent, deliberate evolution process could improve the spec over time without hard breaking changes, and without requiring registries and clients to adopt new versions.

Luckily, the OCI innovation process has _lots_ of experience being _very_ diligent and deliberate, only it's _so_ slow that it's hardly evolved at all. 😅 

> Non-container image tooling must refer to the image-spec

Again, I think this becomes just an unfortunate and mostly invisible misnomer.

The thing we've previously called an "image" is just a descriptor + zero or more layer blob references in a meaningful order. People are already successfully building lots of useful software inside that definition, and I don't see many people eager to stop and build on a new spec instead, just to remove the term "image" from their codebase.

_(Meta point: Is this really the best, most visible forum to be having this discussion? We should make sure all interested parties are involved, and a new 1500-word discussion in the artifacts repo might not be the best place to advertise this conversation, to get the most diverse set of responses -- maybe a TOB issue, with relevant TOB members cc'ed?)_

**Reply by SteveLasker** [reply]
- **existing usage, as outlined in OCI Artifacts**  
We have had great success with a subset of types. However, from the beginning, we had constraints that couldn't support CNAB, or even what Helm would like in the full extent. Helm has the concept of dependencies on images. That's not surfaced in the manifest, so copying a helm chart, and its image dependencies isn't possible in a generic way. This is why we're proposing building upon what's possible today. We don't support search or rich metadata. See here for some goals of [OCI Artifacts and a View of the Future](https://stevelasker.blog/2019/08/25/oci-artifacts-and-a-view-of-the-future/).
- **converting config and layers to optional**  
Changing these to optional requires changes to registries that already have validation enforcing the required nature. Thus, we're back to versioning. A 1.1 release could declare these as optional, but that is a versioned change.
- **breaking changes and expectations**  
This is the crux of the issue. Simply adding an optional property doesn't convey the expectation of a user. Just because a registry or even a client could accept or ignore some random element in an existing schema; what is the expectation of users and the contractual support they have with those products and services? Can a paying customer open a support ticket on a product because OCI decided to add an optional property to a versioned spec?  
A passive annotation that has no expectation of processing by a registry is a non-issue, IMO. Concepts, where there is an expectation of a registry or client to process does suggest an agreement through a version. In the refType case, lifecycle management. When a paying customer can't delete all the content from their registry, reducing their storage costs, because a bunch of reference types were added, with no ability to delete them, who's accountable? 
PR37 attempts to add extensibility for manifests, where there should be a means of negotiation between a client and a registry. We're trying to do the same thing with distribution, as refTypes outlines a `/referrers` API that doesn't yet exist in the distribution spec. 
- **Alternatives**  
We've spent a LOT of time trying to make this new model work. We've looked at various alternatives, prototyped several we felt were viable, spent a lot of time explaining the approach, and this is where we've landed. I don't exactly enjoy these endless debates, and most folks have gone to private discussions and feedback, while I continue to push for publically communicated validated change. We have more to do, and we do expect 29 to evolve into 37, *over time*. We've incorporated every bit of feedback we felt we could, churning back and forth multiple times. while adhering to the goals of a long-lived, evolving product. If someone would like to incubate alternatives, outlining the goals, building functional prototypes that could work across partners and competitors, getting usage and feedback, that would be great.

**Reply by imjasonh** [reply]
> Changing these to optional requires changes to registries that already have validation enforcing the required nature. Thus, we're back to versioning. A 1.1 release could declare these as optional, but that is a versioned change.

If the spec changed to "registries MAY require `config` and `layers`", existing registries that require them and new implementations that don't would both be compliant with the same text, and no new spec version would be required. This is exactly the definition of a backward-compatible change, loosening the requirements on existing compliant implementations while allowing other implementations to enforce other constraints.

> We've incorporated every bit of feedback we felt we could

I don't feel that the feedback that "it's possible to make backward-compatible changes to existing specs, without versioning" has been incorporated or heard, and I have yet to hear a convincing argument against it. You seem to hold this belief despite it actually making your job a _lot_ harder, and requiring a totally new type to be slowly rolled out across existing registries.

For example, https://github.com/opencontainers/image-spec/issues/827 and https://github.com/opencontainers/image-spec/pull/826 are designed to be a backward-compatible change that registries and clients can adopt (or not) on their own timelines, while remaining image-spec compliant, because the spec is designed to be flexible, not frozen in time as soon as 2+ registries implement it.

**Reply by SteveLasker** [reply]
Let's say the image-spec changes the `config` property from REQUIRED to OPTIONAL under the 1.0 spec. A paying customer uses the `foo` client to push a new artifact to the registry that doesn't have a `config`. The registry fails to accept the artifact. The customer opens a support ticket. The customer points to a PR under the image-spec 1.0, saying it was changed to OPTIONAL. 
What is the expected outcome?

**Reply by imjasonh** [reply]
"Thank you for being a valued customer of RegistryCorp™️ International. As you found, the spec was updated to state that registries MAY require `config`, and though some registries support config-less images today, we here at RC have decided to keep it as required for now. Your feedback is important to us, and if you can describe your use case in more detail we can use that information to prioritize config-less images, and possibly provide a timeline for supporting them in an upcoming release of RegistryCorp™️. In the meantime you can use `foo` client v1.2.3 which pushes images with an empty dummy config to satisfy legacy registries that require them."

**Reply by dlorenc** [reply]
Steve - I'm not sure if you intend this or not, but your usage of "we" when describing decisions here appears to convey some kind of consensus decision on behalf of a group, that I don't think exists. It would be much easier to read these if it were made more precise. Is "we" you? Is it your company? Is it a specific group of people? If so - who?

**Reply by SteveLasker** [reply]
@imjasonh:  What is your definition of versioning?

@dlorenc, "we" represents the group of folks that are working on implementing reference types and notary v2 which includes Azure folks, and folks from several other companies. It's not my place to say their names if they choose not to speak publicly, rather provide private feedback as they choose not to be involved in these long debates. I am very clear when it's a me, vs. a we.

**Reply by dlorenc** [reply]
So "we" is "the notary v2" community?

**Reply by mikebrow** [reply]
In the context of the bigger we.. I think it's fair to propose additional schema/artifact patterns that are not a direct copy of the v1 OCI image spec, and if those proposed changes show good merit and adoption I would expect the we to grow in size.  I look at this as a proposal, that is a direct output from the aforementioned workgroup. A proposal to be discussed, not a directive being forced on the unwilling.

**Reply by dlorenc** [reply]
I agree! I just think it's a good idea to qualify the word "we", especially for official members of the OCI TOB (or really any TOB).

We (see what I did there) all wear multiple hats (individual, company, open source community, open source TOB...), and it's important to be precise.

There are multiple options being proposed still (including one of my own), and I speak on behalf of myself here.

**Reply by mikebrow** [reply]
> I agree! I just think it's a good idea to qualify the word "we", especially for official members of the OCI TOB (or really any TOB).
> 
> We (see what I did there) all wear multiple hats (individual, company, open source community, open source TOB...), and it's important to be precise.
> 
> There are multiple options being proposed still (including one of my own), and I speak on behalf of myself here.

Great point.. we agree ;-)

**Reply by SteveLasker** [reply]
I've qualified all _we's_ as these aren't unique to just notary, but a combination of folks working on the reference types and notary scenarios.

**Reply by dlorenc** [reply]
As long as it's clear "we" isn't any officially sanctioned OCI group, I'm happy. It sounds like that's the case?

**Reply by brackendawson** [reply]
To me the GC implications are a clear indication that artifacts should be a new manifest.

If they were an optional part of an image manifest, what does "optional" mean to the implementer of a registry? I can ignore it, but do I store the manifest I receive or the one I decode? Would it not be easy to get a situation where a registry user can successfully use artifacts on a registry which does not support them right up until the artifacts are hit by GC.

**Comment by mikebrow** [comment]
great discussion.. heads up this is going read only on repo archive...
